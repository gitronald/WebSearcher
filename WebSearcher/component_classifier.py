"""SERP component classifiers
"""

from . import webutils

def classify_type(cmpt):
    """Component classifier

    Args:
        cmpt (bs4 object): A Search component

    Returns:
        str: A classification of the component type
    """
    
    # Default unknown
    cmpt_type = "unknown"
    g_tray = cmpt.find("g-tray-header")
    g_section = cmpt.find("g-section-with-header")
    carousel = cmpt.find("g-scrolling-carousel")
    g_accordian = cmpt.find("g-accordion")
    related_question_pair = cmpt.find("related-question-pair")
    knowledge = cmpt.find("div", {"class": ["knowledge-panel", "knavi", "kp-blk"]})
    finance = cmpt.find("div", {"id": "knowledge-finance-wholepage__entity-summary"})
    img_box = cmpt.find("div", {"id": "imagebox_bigimages"})
    hybrid = cmpt.find("div", {"class": "ifM9O"})
    twitter = cmpt.find_previous().text == "Twitter Results"


    # Checks a g-scrolling-carousel for a specific id to classify as not all 
    # top_stories have an h3 tag
    if carousel:
        if webutils.check_dict_value(cmpt.attrs, "class", ["F8yfEe"]):
            cmpt_type = "top_stories"

    if cmpt_type == "unknown":
        cmpt_type = classify_header(cmpt, level=2)
        
    if cmpt_type == "unknown":
        cmpt_type = classify_header(cmpt, level=3)

    if cmpt_type == "unknown" and "class" in cmpt.attrs:
        if any(s in ["hlcw0c", "MjjYud"] for s in cmpt.attrs["class"]):
            cmpt_type = "general"

            if cmpt.find("block-component"):
                # this can also be a "related results box"
                # Check for image card block
                cmpt_type = "img_cards"

    # Twitter subtype
    if twitter or cmpt_type == "twitter":
        cmpt_type = "twitter_cards" if carousel else "twitter_result"

    # Check for binary match only divs (exists/doesn't exist)
    if cmpt_type == "unknown":
        if img_box:
            cmpt_type = "images"
        elif knowledge:
            cmpt_type = "knowledge"
        elif finance:
            cmpt_type = "knowledge"
        elif hybrid and g_accordian:
            cmpt_type = "general_questions"

    # Check for available on divs
    if "/Available on" in cmpt.text:
        cmpt_type = "available_on"

    # Check if component is only of class 'g'
    if webutils.check_dict_value(cmpt.attrs, "class", ["g"]):
        cmpt_type = "general"

    # check for people also ask
    if cmpt_type == "unknown":
        cmpt_type = classify_people_also_ask(cmpt)

    # check for flights, maps, hotels, events, jobs
    if cmpt_type == "unknown":
        cmpt_type = classify_knowledge_box(cmpt)

    # Check for hidden components
    condition = webutils.check_dict_value(cmpt.attrs, "class", ["ULSxyf"])
    if cmpt_type == "unknown" and condition:

        if cmpt.find('promo-throttler'):
            cmpt_type = "hidden-survey"
        
        elif cmpt.find('block-component'):
            cmpt_type = 'knowledge'
    
    # Return type or unknown (default)
    return cmpt_type


def classify_header(cmpt, level):
    """Check text in common headers for dict matches"""

    # Find headers
    if level == 2:
        header_dict = {
            'Calculator Result': 'knowledge',
            'Directions': 'directions',
            'Featured snippet from the web': 'knowledge',
            'Finance Results': 'knowledge',
            'Jobs': 'jobs',
            'Knowledge Result': 'knowledge',
            'Local Results': 'local_results',
            'Map Results': 'map_results',
            'People also ask': 'people_also_ask',
            'Perspectives & opinions': 'perspectives',
            'Perspectives': 'perspectives',
            'Related searches': 'searches_related',
            'Resultado de traducci√≥n': 'knowledge',
            'Resultados de la Web': 'general',
            'Sports Results': 'knowledge',
            'Top stories': 'top_stories',
            'Translation Result': 'knowledge',
            'Twitter Results': 'twitter',
            'Unit Converter': 'knowledge',
            'Weather Result': 'knowledge',
            'Web Result with Site Links': 'general',
            'Web results': 'general',
            'Complementary Results': 'knowledge',
            'Videos': 'videos',
        }
    elif level == 3:
        header_dict = {
            'Images for': 'images',
            'Latest from': 'latest_from',
            'Popular products': 'products',
            'Quotes in the news': 'news_quotes',
            'Recipes': 'recipes',
            'Related searches': 'searches_related',
            'Scholarly articles for': 'scholarly_articles',
            'Top stories': 'top_stories',
            'Videos': 'videos',
            'View more news': 'view_more_news',
            'View more videos': 'view_more_videos'
        }

    # Find headers, eg for level 2: <h2> or <div aria-level="2" role="heading">
    header_list = [
        cmpt.find(f"h{level}"),
        cmpt.find("div", {'aria-level':f"{level}", "role":"heading"})
    ]

   # Check for string matches in header text e.g. `h2.text`
    for header in filter(None, header_list):
        for text, label in header_dict.items():
            if header.text.startswith(text):
                return label

    # Return unknown if no matches
    return "unknown"


def classify_people_also_ask(cmpt):
    """Secondary check for people also ask, see classify_header for primary"""
    class_list = ["g", "kno-kp", "mnr-c", "g-blk"]
    conditions = webutils.check_dict_value(cmpt.attrs, "class", class_list)
    return 'people_also_ask' if conditions else "unknown"


def classify_knowledge_box(cmpt):
    """Classify knowledge component types
    
    Creates conditions for each label in a dictionary then assigns the 
    label as the component type if it's conditions are met using one, or some 
    combination, of the following three methods:

    1. Check if a component has a specific attribute value
    2. Check if a component contains a specific div
    3. Check if a component has a matching string in its text
    
    """
    attrs = cmpt.attrs

    condition = {}
    condition['flights'] = (
        (webutils.check_dict_value(attrs, "jscontroller", "Z2bSc")) |
        bool(cmpt.find("div", {"jscontroller": "Z2bSc"}))
    )
    condition['maps'] = webutils.check_dict_value(attrs, "data-hveid", "CAMQAA")
    condition['hotels'] = cmpt.find("div", {"class": "zd2Jbb"})
    condition['events'] = cmpt.find("g-card", {"class": "URhAHe"})
    condition['jobs'] = cmpt.find("g-card", {"class": "cvoI5e"})

    text_list = list(cmpt.stripped_strings)
    if text_list:
        condition['covid_alert'] = (text_list[0] == "COVID-19 alert")

    for condition_type, conditions in condition.items():
        return condition_type if conditions else "unknown"
