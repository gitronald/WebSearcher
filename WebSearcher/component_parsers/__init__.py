import pandas as pd
from .ads import parse_ads
from .knowledge import parse_knowledge_panel
from .general import parse_general_results
from .top_stories import parse_top_stories
from .latest_from import parse_latest_from
from .view_more_news import parse_view_more_news
from .news_quotes import parse_news_quotes
from .people_also_ask import parse_people_also_ask
from .local_results import parse_local_results
from .map_results import parse_map_results
from .images import parse_images
from .videos import parse_videos
from .scholarly_articles import parse_scholarly_articles
from .twitter_cards import parse_twitter_cards
from .twitter_result import parse_twitter_result
from .general_questions import parse_general_questions
from .available_on import parse_available_on
from .footer import parse_footer

# Component details dataframe
columns = ['type', 'func', 'label']
components = pd.DataFrame([
    ('ad', parse_ads, 'Ad'),
    ('knowledge', parse_knowledge_panel, 'Knowledge'),
    ('general', parse_general_results, 'General'),
    ('general_questions', parse_general_questions, 'General Questions'),
    ('general_menu', parse_general_results, 'General Submenu'),
    ('available_on', parse_available_on, 'Available On'),
    ('top_stories', parse_top_stories, 'Top Stories'), 
    ('latest_from', parse_latest_from, 'Latest From'),
    ('view_more_news', parse_view_more_news, 'View More News'),
    ('news_quotes', parse_news_quotes, 'News Quotes'),
    ('people_also_ask', parse_people_also_ask, 'People Also Ask'),
    ('local_results', parse_local_results, 'Local Results'),
    ('map_results', parse_map_results, 'Map Results'),
    ('images', parse_images, 'Images'),
    ('videos', parse_videos, 'Videos'),
    ('view_more_videos', parse_videos, 'View More Videos'),
    ('twitter_cards', parse_twitter_cards, 'Twitter Cards'),
    ('twitter_result', parse_twitter_result, 'Twitter Result'),
    ('scholarly_articles', parse_scholarly_articles, 'Scholar Articles'),
    ('footer', parse_footer, 'Footer')
], columns=columns)

# dict[type] = function
type_functions = components.set_index('type')['func'].to_dict()

# dict[type] = label
type_labels = components.set_index('type')['label'].to_dict()
