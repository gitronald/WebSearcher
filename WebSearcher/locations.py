import os
import base64
import string
import requests
import pandas as pd
from bs4 import BeautifulSoup

from . import logger
log = logger.Logger().start(__name__)

url = 'https://developers.google.com/adwords/api/docs/appendix/geotargeting'

def get_all_urls(soup):
    a_divs = soup.find_all('a')
    all_urls = {a.attrs['href'] for a in a_divs if 'href' in a.attrs}    
    return all_urls

def download_locations(data_dir, url=url, return_data=True):
    """Download the latest locations data

    Checks if the current version already exists locally before downloading
    
    Args:
        data_dir (str): Where to save the data as a csv
        url (str, optional): Defaults to the current URL
    
    Raises:
        SystemExit: Exit if file exists

    Returns:
        None: Saves to file in the default or selected data_dir

    """
    data_dir = data_dir if data_dir else 'data/locations'
    os.makedirs(data_dir, exist_ok=True)

    # Get latest URL
    try:
        html = requests.get(url).content
        soup = BeautifulSoup(html, 'lxml')
        all_urls = get_all_urls(soup)
        geo_urls = [url for url in all_urls if 'geotargets' in url]

    except Exception:
        log.exception("Failed to retrieve location data's url")

    # Get current CSV url and use as filename
    geo_url = sorted(geo_urls)[-1]
    full_url = 'https://developers.google.com' + geo_url
    fp = os.path.join(data_dir, geo_url.split('/')[-1])

    # Check if the current version already exists
    if os.path.exists(fp):
        print(f'Version up to date: {geo_url}')
    else:
        # If it doesn't, download it
        try:
            print(f'Getting: {full_url}')
            locations = pd.read_csv(full_url)
        except Exception:
            log.exception('Failed to retrieve location data')

        # Save
        print(f"Saving: {fp}")
        locations.to_csv(fp, index=False, encoding='utf-8')
        
        # Return
        if return_data:
            return locations


def get_location_id(canonical_name):
    """Get location ID for URL parameter 'uule'
    
    Returns the url parameter for a given location's Canonical Name
    
    Args:
        canonical_name (str): Canoncial Name for a location, see
        data downloaded using download_locations. Column name is 
        usually something like "Canonical Name" or "Canonical.Name". 
    
    Returns:
        str: The parameter key for selecting a location
    
    """
    uule_key = string.ascii_uppercase+string.ascii_lowercase+string.digits
    uule_key = uule_key + '-_' + uule_key + '-_' # Double length, repeating
    key = uule_key[len(canonical_name)]
    b64 = base64.b64encode(canonical_name.encode('utf-8')).decode('utf-8')
    return f'w+CAIQICI{key}{b64}'