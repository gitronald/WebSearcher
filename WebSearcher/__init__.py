__version__ = "2023.05.19"
from .searchers import SearchEngine
from .parsers import parse_serp, extract_components
from .locations import download_locations
from .component_classifier import classify_type
from .webutils import load_html, make_soup, load_soup
