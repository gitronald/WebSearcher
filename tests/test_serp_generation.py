import pytest
import glob
import WebSearcher as ws

from syrupy.extensions.json import JSONSnapshotExtension

@pytest.fixture
def snapshot_json(snapshot):
    return snapshot.use_extension(JSONSnapshotExtension)

def pytest_generate_tests(metafunc):
    file_list = glob.glob('./tests/html_pages/*.html')
    metafunc.parametrize("file_name", file_list )

def test_parsing(snapshot_json, file_name):
    # read html
    with open(file_name) as file:
        html = file.read()
    
    # Initialize crawler
    se = ws.SearchEngine()
    
    # Conduct Search
    se.mock_search(html)

    # Parse Results
    se.parse_results()

    assert se.results == snapshot_json
