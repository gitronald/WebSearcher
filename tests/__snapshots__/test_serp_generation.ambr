# serializer version: 1
# name: test_serp_1684837514_is_correct
  '''
  [
      {
          "type": "top_stories",
          "sub_rank": 0,
          "title": "Texas Forced This Woman to Give Birth  | to a Stillborn Son. She\u2019s Suing | 15 hours ago",
          "url": "https://www.rollingstone.com/politics/politics-features/texas-abortion-ban-forced-birth-1234739485/",
          "cite": null,
          "timestamp": null,
          "details": {
              "img_url": null,
              "orient": "h",
              "live_stamp": false
          },
          "cmpt_rank": 0,
          "serp_rank": 0,
          "crawl_id": null,
          "serp_id": "f468833b92bcba04d43b110f32916a4da64153ba553949ace7e451f7",
          "qry": "abortion",
          "lang": "en",
          "lhs_bar": false
      },
      {
          "type": "top_stories",
          "sub_rank": 1,
          "title": "NPR | More women sue Texas  | saying the state's anti- | abortion laws harmed them | 20 hours ago",
          "url": "https://www.npr.org/2023/05/22/1177425651/texas-women-sue-abortion",
          "cite": null,
          "timestamp": null,
          "details": {
              "img_url": null,
              "orient": "h",
              "live_stamp": false
          },
          "cmpt_rank": 0,
          "serp_rank": 1,
          "crawl_id": null,
          "serp_id": "f468833b92bcba04d43b110f32916a4da64153ba553949ace7e451f7",
          "qry": "abortion",
          "lang": "en",
          "lhs_bar": false
      },
      {
          "type": "top_stories",
          "sub_rank": 2,
          "title": "Yahoo News | More women sue Texas,  | asking court to put  | emergency block on state's \nabortion law | 21 hours ago",
          "url": "https://news.yahoo.com/more-women-sue-texas-asking-123339247.html",
          "cite": null,
          "timestamp": null,
          "details": {
              "img_url": null,
              "orient": "h",
              "live_stamp": false
          },
          "cmpt_rank": 0,
          "serp_rank": 2,
          "crawl_id": null,
          "serp_id": "f468833b92bcba04d43b110f32916a4da64153ba553949ace7e451f7",
          "qry": "abortion",
          "lang": "en",
          "lhs_bar": false
      },
      {
          "type": "top_stories",
          "sub_rank": 3,
          "title": "AP News | Nebraska governor  | signs 12-week abortion  | ban, limits on gender-affirming \ncare for minors | 18 hours ago",
          "url": "https://apnews.com/article/transgender-health-care-abortion-ban-nebraska-1a88067bf403559116df7c09e004e472",
          "cite": null,
          "timestamp": null,
          "details": {
              "img_url": null,
              "orient": "h",
              "live_stamp": false
          },
          "cmpt_rank": 0,
          "serp_rank": 3,
          "crawl_id": null,
          "serp_id": "f468833b92bcba04d43b110f32916a4da64153ba553949ace7e451f7",
          "qry": "abortion",
          "lang": "en",
          "lhs_bar": false
      },
      {
          "type": "top_stories",
          "sub_rank": 4,
          "title": "Oregon Capital Chronicle | Here's what Oregon's  | controversial abortion,  | gender-affirming care bill \nwould do \u2013 Oregon Capital Chronicle | 21 hours ago",
          "url": "https://oregoncapitalchronicle.com/2023/05/22/heres-what-oregons-controversial-abortion-gender-affirming-care-bill-would-do/",
          "cite": null,
          "timestamp": null,
          "details": {
              "img_url": null,
              "orient": "h",
              "live_stamp": false
          },
          "cmpt_rank": 0,
          "serp_rank": 4,
          "crawl_id": null,
          "serp_id": "f468833b92bcba04d43b110f32916a4da64153ba553949ace7e451f7",
          "qry": "abortion",
          "lang": "en",
          "lhs_bar": false
      },
      {
          "type": "top_stories",
          "sub_rank": 5,
          "title": "USA Today | Tuberville abortion  | policy fight threatens US  | security, Pentagon says | 1 hour ago",
          "url": "https://www.usatoday.com/story/news/politics/2023/05/23/senator-tommy-tuberville-abortion-national-security/70232023007/",
          "cite": null,
          "timestamp": null,
          "details": {
              "img_url": null,
              "orient": "h",
              "live_stamp": false
          },
          "cmpt_rank": 0,
          "serp_rank": 5,
          "crawl_id": null,
          "serp_id": "f468833b92bcba04d43b110f32916a4da64153ba553949ace7e451f7",
          "qry": "abortion",
          "lang": "en",
          "lhs_bar": false
      },
      {
          "type": "top_stories",
          "sub_rank": 6,
          "title": "NPR | Doctors are reluctant to  | practice in states that  | banned abortion : Shots - \nHealth News | 1 hour ago",
          "url": "https://www.npr.org/sections/health-shots/2023/05/23/1177542605/abortion-bans-drive-off-doctors-and-put-other-health-care-at-risk",
          "cite": null,
          "timestamp": null,
          "details": {
              "img_url": null,
              "orient": "h",
              "live_stamp": false
          },
          "cmpt_rank": 0,
          "serp_rank": 6,
          "crawl_id": null,
          "serp_id": "f468833b92bcba04d43b110f32916a4da64153ba553949ace7e451f7",
          "qry": "abortion",
          "lang": "en",
          "lhs_bar": false
      },
      {
          "type": "perspectives",
          "sub_rank": 0,
          "title": "Michelle Goldberg | The New York Times | Opinion | You Cannot  | Hear These 13  | Women's Stories and  | Believe the \nAnti-Abortion Narrative | It's increasingly clear that  | it's not safe to be pregnant  | in states with \ntotal abortion bans. Since the end of Roe v. | . | 11 hours ago",
          "url": "https://www.nytimes.com/2023/05/22/opinion/abortion-law-texas-lawsuit.html",
          "cite": null,
          "text": "It's increasingly clear that  it's not safe to be pregnant  in states with \ntotal abortion bans. Since the end of Roe v.",
          "timestamp": null,
          "details": {
              "img_url": null,
              "orient": "h",
              "live_stamp": false
          },
          "cmpt_rank": 1,
          "serp_rank": 7,
          "crawl_id": null,
          "serp_id": "f468833b92bcba04d43b110f32916a4da64153ba553949ace7e451f7",
          "qry": "abortion",
          "lang": "en",
          "lhs_bar": false
      },
      {
          "type": "perspectives",
          "sub_rank": 1,
          "title": "Jill Filipovic | Substack | New Study: Abortion  | Bans Endanger  | Women's Lives | Where abortion is  | outlawed, women receive  | substandard healthcare.  | Some are \nfighting back. | . | 16 hours ago",
          "url": "https://jill.substack.com/p/new-study-abortion-bans-endanger",
          "cite": null,
          "text": "Where abortion is  outlawed, women receive  substandard healthcare.  Some are \nfighting back.",
          "timestamp": null,
          "details": {
              "img_url": null,
              "orient": "h",
              "live_stamp": false
          },
          "cmpt_rank": 1,
          "serp_rank": 8,
          "crawl_id": null,
          "serp_id": "f468833b92bcba04d43b110f32916a4da64153ba553949ace7e451f7",
          "qry": "abortion",
          "lang": "en",
          "lhs_bar": false
      },
      {
          "type": "perspectives",
          "sub_rank": 2,
          "title": "Sarah Green Carmichael, Lisa Jarvis | Bloomberg.com | Yet Another Bad Week  | for Abortion Rights in  | the US | Judges on the 5th Circuit  | appeared worryingly  | comfortable challenging  | the \nexpertise of FDA scientists. | . | 1 day ago",
          "url": "https://www.bloomberg.com/opinion/articles/2023-05-21/yet-another-bad-week-for-abortion-rights-in-the-us",
          "cite": null,
          "text": "Judges on the 5th Circuit  appeared worryingly  comfortable challenging  the \nexpertise of FDA scientists.",
          "timestamp": null,
          "details": {
              "img_url": null,
              "orient": "h",
              "live_stamp": false
          },
          "cmpt_rank": 1,
          "serp_rank": 9,
          "crawl_id": null,
          "serp_id": "f468833b92bcba04d43b110f32916a4da64153ba553949ace7e451f7",
          "qry": "abortion",
          "lang": "en",
          "lhs_bar": false
      },
      {
          "type": "knowledge",
          "sub_rank": 0,
          "subtype": "panel",
          "details": {
              "heading": "People also ask",
              "urls": [
                  {
                      "url": "https://en.wikipedia.org/wiki/Abortion",
                      "text": "Wikipedia"
                  }
              ],
              "text": "Choose what you\u2019re giving feedback on|general feedback|Abortion|Abortion|Abortion is the termination of a pregnancy by removal or expulsion of an embryo or fetus. An abortion that occurs without intervention is known as a miscarriage or \"spontaneous abortion\"; these occur in approximately 30% to 40% of all pregnancies.| |Wikipedia|:|Induced miscarriage, termination of pregnancy|Induced miscarriage, termination of pregnancy|:|10A0|10A0|:|779.6|779.6|Feedback|Feedback|Feedback|Feedback|People also ask|People also ask|People also ask|You will see more English now.|You will see more English now.|You will see more English now.|You will see more English now.|You will see more English now.|What is abortion in medical term?|What is abortion in medical term?|What is abortion in medical term?|What is missed abortion?|What is missed abortion?|What is missed abortion?|What is the legal definition of abortion?|What is the legal definition of abortion?|What is the legal definition of abortion?|What are the different classification of abortion?|What are the different classification of abortion?|What are the different classification of abortion?|Feedback",
              "img_url": null
          },
          "cmpt_rank": 2,
          "serp_rank": 10,
          "crawl_id": null,
          "serp_id": "f468833b92bcba04d43b110f32916a4da64153ba553949ace7e451f7",
          "qry": "abortion",
          "lang": "en",
          "lhs_bar": false
      },
      {
          "type": "general",
          "sub_rank": 0,
          "title": "Abortion",
          "url": "https://www.who.int/news-room/fact-sheets/detail/abortion",
          "cite": "https://www.who.int \u203a Newsroom \u203a Fact sheets \u203a Detail",
          "details": "",
          "text": "Nov 25, 2021  \u2014  Abortion  is a common health intervention. It is safe when carried out using a method recommended by WHO, appropriate to the pregnancy\u00a0...",
          "timestamp": "Nov 25, 2021",
          "cmpt_rank": 3,
          "serp_rank": 11,
          "crawl_id": null,
          "serp_id": "f468833b92bcba04d43b110f32916a4da64153ba553949ace7e451f7",
          "qry": "abortion",
          "lang": "en",
          "lhs_bar": false
      },
      {
          "type": "local_results",
          "sub_rank": 0,
          "title": "Boulder Valley Women's Health Center",
          "details": {
              "website": "http://www.boulderwomenshealth.org/",
              "directions": "#"
          },
          "url": "http://www.boulderwomenshealth.org/",
          "cmpt_rank": 4,
          "serp_rank": 12,
          "crawl_id": null,
          "serp_id": "f468833b92bcba04d43b110f32916a4da64153ba553949ace7e451f7",
          "qry": "abortion",
          "lang": "en",
          "lhs_bar": false
      },
      {
          "type": "local_results",
          "sub_rank": 1,
          "title": "Planned Parenthood - Boulder",
          "details": {
              "website": "https://www.plannedparenthood.org/health-center/colorado/boulder/80302/boulder-2488-90210?utm_campaign=boulder&utm_medium=organic&utm_source=local-listing",
              "directions": "#"
          },
          "url": "https://www.plannedparenthood.org/health-center/colorado/boulder/80302/boulder-2488-90210?utm_campaign=boulder&utm_medium=organic&utm_source=local-listing",
          "cmpt_rank": 4,
          "serp_rank": 13,
          "crawl_id": null,
          "serp_id": "f468833b92bcba04d43b110f32916a4da64153ba553949ace7e451f7",
          "qry": "abortion",
          "lang": "en",
          "lhs_bar": false
      },
      {
          "type": "local_results",
          "sub_rank": 2,
          "title": "Boulder Abortion Clinic",
          "details": {
              "website": "http://www.drhern.com/",
              "directions": "#"
          },
          "url": "http://www.drhern.com/",
          "cmpt_rank": 4,
          "serp_rank": 14,
          "crawl_id": null,
          "serp_id": "f468833b92bcba04d43b110f32916a4da64153ba553949ace7e451f7",
          "qry": "abortion",
          "lang": "en",
          "lhs_bar": false
      },
      {
          "type": "general",
          "sub_rank": 0,
          "title": "Abortion Information | Information About Your Options",
          "url": "https://www.plannedparenthood.org/learn/abortion",
          "cite": "https://www.plannedparenthood.org \u203a learn \u203a abortion",
          "details": "",
          "text": "Abortion  is a safe and legal way to end a pregnancy. Learning more about  abortion  methods and options can help you decide if it is the right choice for you.",
          "timestamp": null,
          "cmpt_rank": 5,
          "serp_rank": 15,
          "crawl_id": null,
          "serp_id": "f468833b92bcba04d43b110f32916a4da64153ba553949ace7e451f7",
          "qry": "abortion",
          "lang": "en",
          "lhs_bar": false
      },
      {
          "type": "general",
          "sub_rank": 0,
          "title": "What is Abortion?",
          "url": "https://www.bpas.org/abortion-care/considering-abortion/what-is-abortion/",
          "cite": "https://www.bpas.org \u203a considering-abortion \u203a what-is...",
          "details": "",
          "text": "An  abortion  is when the pregnancy is ended so that it does not result in the birth of a child. Sometimes this is called 'termination of pregnancy'.",
          "timestamp": null,
          "cmpt_rank": 6,
          "serp_rank": 16,
          "crawl_id": null,
          "serp_id": "f468833b92bcba04d43b110f32916a4da64153ba553949ace7e451f7",
          "qry": "abortion",
          "lang": "en",
          "lhs_bar": false
      },
      {
          "type": "general",
          "sub_rank": 0,
          "title": "Abortion",
          "url": "https://en.wikipedia.org/wiki/Abortion",
          "cite": "https://en.wikipedia.org \u203a wiki \u203a Abortion",
          "details": "",
          "text": "An  abortion  that occurs without intervention is known as a miscarriage or \"spontaneous  abortion \"; these occur in approximately 30% to 40% of all pregnancies.",
          "timestamp": null,
          "cmpt_rank": 7,
          "serp_rank": 17,
          "crawl_id": null,
          "serp_id": "f468833b92bcba04d43b110f32916a4da64153ba553949ace7e451f7",
          "qry": "abortion",
          "lang": "en",
          "lhs_bar": false
      },
      {
          "type": "general",
          "sub_rank": 0,
          "title": "The World's Abortion Laws",
          "url": "https://reproductiverights.org/maps/worlds-abortion-laws/",
          "cite": "https://reproductiverights.org \u203a Maps",
          "details": "",
          "text": "The World  Abortion  Laws Map is the definitive record of the legal status of  abortion  in countries across the globe. Since 1998, the Center for.",
          "timestamp": null,
          "cmpt_rank": 8,
          "serp_rank": 18,
          "crawl_id": null,
          "serp_id": "f468833b92bcba04d43b110f32916a4da64153ba553949ace7e451f7",
          "qry": "abortion",
          "lang": "en",
          "lhs_bar": false
      },
      {
          "type": "videos",
          "sub_rank": 0,
          "url": "https://www.youtube.com/watch?v=IJHkjZfXMNY",
          "title": "Rep. Deborah Ross: North Carolina abortion ban turns the ...YouTube \u00b7 MSNBC6 minutes, 10 seconds1 day ago",
          "cite": null,
          "timestamp": null,
          "details": {
              "img_url": null,
              "key_moments": false
          },
          "cmpt_rank": 9,
          "serp_rank": 19,
          "crawl_id": null,
          "serp_id": "f468833b92bcba04d43b110f32916a4da64153ba553949ace7e451f7",
          "qry": "abortion",
          "lang": "en",
          "lhs_bar": false
      },
      {
          "type": "videos",
          "sub_rank": 1,
          "url": "https://www.youtube.com/watch?v=g-BqlQegH5Q",
          "title": "Byron Donalds sidesteps his stance on six-week abortion ban ...YouTube \u00b7 NBC News1 minute, 45 seconds1 day ago",
          "cite": null,
          "timestamp": null,
          "details": {
              "img_url": null,
              "key_moments": false
          },
          "cmpt_rank": 9,
          "serp_rank": 20,
          "crawl_id": null,
          "serp_id": "f468833b92bcba04d43b110f32916a4da64153ba553949ace7e451f7",
          "qry": "abortion",
          "lang": "en",
          "lhs_bar": false
      },
      {
          "type": "videos",
          "sub_rank": 2,
          "url": "https://www.youtube.com/watch?v=Xgq-NMPqsbY",
          "title": "Full Panel: Biden is 'testing the waters' using abortion issue to ...YouTube \u00b7 NBC News7 minutes, 46 seconds1 day ago",
          "cite": null,
          "timestamp": null,
          "details": {
              "img_url": null,
              "key_moments": false
          },
          "cmpt_rank": 9,
          "serp_rank": 21,
          "crawl_id": null,
          "serp_id": "f468833b92bcba04d43b110f32916a4da64153ba553949ace7e451f7",
          "qry": "abortion",
          "lang": "en",
          "lhs_bar": false
      },
      {
          "type": "general",
          "sub_rank": 0,
          "title": "Medical Abortion",
          "url": "https://medlineplus.gov/abortion.html",
          "cite": "https://medlineplus.gov \u203a Health Topics",
          "details": "",
          "text": "Aug 5, 2022  \u2014  An  abortion  is a medical procedure to end a pregnancy. It uses medicine or surgery to remove the embryo or fetus and placenta from the\u00a0...",
          "timestamp": "Aug 5, 2022",
          "cmpt_rank": 10,
          "serp_rank": 22,
          "crawl_id": null,
          "serp_id": "f468833b92bcba04d43b110f32916a4da64153ba553949ace7e451f7",
          "qry": "abortion",
          "lang": "en",
          "lhs_bar": false
      },
      {
          "type": "general",
          "sub_rank": 0,
          "title": "CDCs Abortion Surveillance System FAQs",
          "url": "https://www.cdc.gov/reproductivehealth/data_stats/abortion.htm",
          "cite": "https://www.cdc.gov \u203a reproductivehealth \u203a data_stats",
          "details": "",
          "text": "Among 48 reporting areas with data each year during 2011\u20132020, in 2020, a total of 615,911 abortions were reported, the  abortion  rate was 11.2 abortions per\u00a0...",
          "timestamp": null,
          "cmpt_rank": 11,
          "serp_rank": 23,
          "crawl_id": null,
          "serp_id": "f468833b92bcba04d43b110f32916a4da64153ba553949ace7e451f7",
          "qry": "abortion",
          "lang": "en",
          "lhs_bar": false
      },
      {
          "type": "people_also_ask",
          "sub_rank": 0,
          "details": null,
          "cmpt_rank": 12,
          "serp_rank": 24,
          "crawl_id": null,
          "serp_id": "f468833b92bcba04d43b110f32916a4da64153ba553949ace7e451f7",
          "qry": "abortion",
          "lang": "en",
          "lhs_bar": false
      },
      {
          "type": "searches_related",
          "sub_rank": 1,
          "details": [
              {
                  "text": "long-term effects of unsafe abortion",
                  "url": "/search?q=long-term+effects+of+unsafe+abortion&sa=X&ved=2ahUKEwiT54ngnIv_AhX1lGoFHSMFB80Q1QJ6BAhuEAE"
              },
              {
                  "text": "causes of abortion in teenage",
                  "url": "/search?q=Causes+of+abortion+in+teenage&sa=X&ved=2ahUKEwiT54ngnIv_AhX1lGoFHSMFB80Q1QJ6BAhtEAE"
              },
              {
                  "text": "7 types of abortion ppt",
                  "url": "/search?q=7+types+of+abortion+ppt&sa=X&ved=2ahUKEwiT54ngnIv_AhX1lGoFHSMFB80Q1QJ6BAhqEAE"
              },
              {
                  "text": "types of abortion nursing",
                  "url": "/search?q=Types+of+abortion+nursing&sa=X&ved=2ahUKEwiT54ngnIv_AhX1lGoFHSMFB80Q1QJ6BAhmEAE"
              },
              {
                  "text": "what is abortion brainly",
                  "url": "/search?q=What+is+abortion+Brainly&sa=X&ved=2ahUKEwiT54ngnIv_AhX1lGoFHSMFB80Q1QJ6BAhjEAE"
              },
              {
                  "text": "safe abortion care",
                  "url": "/search?q=Safe+abortion+care&sa=X&ved=2ahUKEwiT54ngnIv_AhX1lGoFHSMFB80Q1QJ6BAhgEAE"
              },
              {
                  "text": "importance of abortion in public health",
                  "url": "/search?q=Importance+of+abortion+in+public+health&sa=X&ved=2ahUKEwiT54ngnIv_AhX1lGoFHSMFB80Q1QJ6BAhcEAE"
              }
          ],
          "cmpt_rank": 12,
          "serp_rank": 25,
          "crawl_id": null,
          "serp_id": "f468833b92bcba04d43b110f32916a4da64153ba553949ace7e451f7",
          "qry": "abortion",
          "lang": "en",
          "lhs_bar": false
      },
      {
          "type": "knowledge",
          "subtype": "panel_rhs",
          "sub_rank": 0,
          "title": "",
          "subtitle": "",
          "text": "Abortion is the termination of a pregnancy by removal or expulsion of an embryo or fetus. An abortion that occurs without intervention is known as a miscarriage or \"spontaneous abortion\"; these occur in approximately 30% to 40% of all pregnancies.",
          "url": "https://en.wikipedia.org/wiki/Abortion",
          "img_urls": [],
          "details": null,
          "cmpt_rank": 13,
          "serp_rank": 26,
          "crawl_id": null,
          "serp_id": "f468833b92bcba04d43b110f32916a4da64153ba553949ace7e451f7",
          "qry": "abortion",
          "lang": "en",
          "lhs_bar": false
      }
  ]
  '''
# ---
